### ISSUE TITLE
Brainstorming a more general API

### ISSUE THREAD

**Issue by juliohm** [xukai92/DensityRatioEstimation.jl#15]
The general idea of this thread is to agree on a general API that scales with large datasets. In such API we should separate well method parameters from the actual abstract problem of density ration estimation. I've finished reading the book by Sugiyama, and I have a clearer picture now of all the methods that are available.

The first thing I would like to propose here is a change to the main function:

```Julia
estimate_ratio(x_nu, x_de, dre)
```

We should be able to handle any iterators `x_nu` and `x_de` of samples, and be able to specify any density ratio estimator `dre`. One thing that we need to decide in terms of the API is whether we name these estimators differently for each optimisation package as discussed in #11, or we add an additional argument to the call to let the user choose the optimization package:

```Julia
estimate_ratio(x_nu, x_de, dre; opt=Optim) = estimate_ratio_impl(x_nu, x_de, dre, opt)
```

I think this latter option is more general, and will avoid us having to create strange names for methods like MMDAnalytical vs MMDJump or something. What do you think?

With the signature of this main function defined, we then need to organise the source code with Requires.jl such that each implementation lives in its own file that is loaded whenever the optimization package is loaded. Analytical or pure Julia implementations could be loaded regardless of the presence of an optimization package, but they would still live in their own separate files.

I have two implementations of KLIEP that I would like to add. One is based on Optim.jl which I think we could consider the default package (pure Julia). And another implementation based on Convex.jl. The question here is if we choose Optim.jl as the default package and add it as a formal dependency or if we don't give him this special status, and treat it as any other optimization package.

Now regarding the iterators `x_ne` and `x_de`, we can write all estimators assuming that we have one sample at a time, and then write fallback methods for pre-allocated Julia matrices. So in this sense, we would rewrite the existing MMD implementations to not allocate memory but simply build the Gramian matrices with a loop on the iterators. Maybe we can assume that these iterators are also indexable, but I am not sure, that may constrain the use cases.

The result of the `estimate_ratio` needs to be designed as well. It can be a weight function that takes new samples and produces weights, or it can be a simple list of weights for the data set passed. Some methods do not support the calculation of weights to points outside the data. So in that sense returning a list of weights is more general. However, it is useful to evaluate the weight function in other points when it is possible. Maybe we should create an abstract result type for these weights, and subtype it with the two scenarios. Some methods will return one type of weight and some methods will return another type.

What do you think of this design? I can start working on it after I finish some research with KLIEP.

**Comment by juliohm** [comment]
@xukai92 did you have a chance to take a look into this?

**Comment by xukai92** [comment]
Sorry for my late response @juliohm, I was on holiday in the past few days.

> We should be able to handle any iterators x_nu and x_de of samples, and be able to specify any density ratio estimator dre. 

It makes sense to me.

> I think this latter option is more general, and will avoid us having to create strange names for methods like MMDAnalytical vs MMDJump or something. What do you think?

I like your proposal as well! Let's stick to it.

> I have two implementations of KLIEP that I would like to add. One is based on Optim.jl which I think we could consider the default package (pure Julia). And another implementation based on Convex.jl. The question here is if we choose Optim.jl as the default package and add it as a formal dependency or if we don't give him this special status, and treat it as any other optimization package.

> Now regarding the iterators x_ne and x_de, we can write all estimators assuming that we have one sample at a time, and then write fallback methods for pre-allocated Julia matrices. 

Exactly!

> So in this sense, we would rewrite the existing MMD implementations to not allocate memory but simply build the Gramian matrices with a loop on the iterators. Maybe we can assume that these iterators are also indexable, but I am not sure, that may constrain the use cases.

I'm not sure how to compute the required Gram matrix through iterators. We can certaintly compute Gram matrices for each batch of data in each iteration. But I don't think those matrices can be used to compute the Gram matrix of the whole dataset. However, if we only iterate x_de, it's OK. Do I miss something here?

> The result of the estimate_ratio needs to be designed as well. It can be a weight function that takes new samples and produces weights, or it can be a simple list of weights for the data set passed. Some methods do not support the calculation of weights to points outside the data. So in that sense returning a list of weights is more general. However, it is useful to evaluate the weight function in other points when it is possible. Maybe we should create an abstract result type for these weights, and subtype it with the two scenarios. Some methods will return one type of weight and some methods will return another type.

I guess there are two possible returns: a vector or a function. Thinking about this, the function `estimate_ratio` kind of implies that it returns a vector of weights. Maybe we should have another function named `fit_ratio_estimator` for those who returns a function?

**Comment by juliohm** [comment]
> I'm not sure how to compute the required Gram matrix through iterators. We can certaintly compute Gram matrices for each batch of data in each iteration. But I don't think those matrices can be used to compute the Gram matrix of the whole dataset. However, if we only iterate x_de, it's OK. Do I miss something here?

I think we are talking the same language :) We just need to iterate over the samples to construct the Gram matrix. My question was more about if we should also assume that the samples are indexable. I think we should not assume that, just iterable.

> I guess there are two possible returns: a vector or a function. Thinking about this, the function `estimate_ratio` kind of implies that it returns a vector of weights. Maybe we should have another function named `fit_ratio_estimator` for those who returns a function?

Let's think about it more broadly after we have an initial implementation. My first impression is that it would be enough to have a single function name `estimate_ratio` that returns a weight type. One weight type will be callable and another weight type will just be iterable.

I will try to work on this next year. December is full of deadlines here for article submissions, and I will hardly be able to touch the code until January.

**Comment by xukai92** [comment]
> I think we should not assume that, just iterable.

I agree. Iterable should be enough.

> One weight type will be callable and another weight type will just be iterable.

I think this works. My concern here is not about the returned type but rather if the function name is still appropriate.

> I will try to work on this next year. December is full of deadlines here for article submissions, and I will hardly be able to touch the code until January.

No worries. Any luck you will go to NeurIPS this year? We can meet in person if you will be there.

**Comment by juliohm** [comment]
It would be nice to meet in person! Unfortunately I won't be there. I am targeting some other conferences this year related to geostatistical learning. There is a lot of work to do :)

**Comment by xukai92** [comment]
Good luck with your submissions then :)

**Comment by juliohm** [comment]
Fixed by #16 :+1:
